scalyr_agent/monitor_utils/k8s.py

```diff
@@ -2142,34 +2147,63 @@ def __init__(self, k8s, port=10255, host_ip=None):
         self._session = requests.Session()
         headers = {
             "Accept": "application/json",
+            "Authorization": "Bearer %s" % k8s.token,
         }
         self._session.headers.update(headers)
 
-        global_log.info("KubeletApi host ip = %s" % host_ip)
-        if host_ip:
-            self._http_host = "http://%s:%d" % (host_ip, port)
-        else:
-            self._http_host = None
+        global_log.info("KubeletApi host ip = %s" % self._host_ip)
+        self._build_host()
         self._timeout = 20.0
 
+    def _build_host(self):
+        if self._host_ip:
+            self._host = "%s://%s:%d" % (self._protocol, self._host_ip, self._port)
+        else:
+            self._host = None
+
+    def _switch_to_fallback(self):
+        self._fallback = True
+        self._port = 10255
+        self._protocol = "http"
+        self._build_host()
+
     def query_api(self, path):
         """ Queries the kubelet API at 'path', and converts OK responses to JSON objects
         """
-        url = self._http_host + path
-        response = self._session.get(url, timeout=self._timeout)
+        url = self._host + path
+        try:
+            response = self._session.get(url, timeout=self._timeout, verify=False)
+        except ConnectionError as e:
+            if not self._fallback:
+                global_log.warning(
+                    "Connection error while querying the Kubelet API: %s. Falling back to older endpoint."
+                    % six.text_type(e),
+                )
+                self._switch_to_fallback()
+                return self.query_api(path)
+            else:
+                raise e
         response.encoding = "utf-8"
         if response.status_code != 200:
-            global_log.log(
-                scalyr_logging.DEBUG_LEVEL_3,
-                "Invalid response from Kubelet API.\n\turl: %s\n\tstatus: %d\n\tresponse length: %d"
-                % (url, response.status_code, len(response.text)),
-                limit_once_per_x_secs=300,
-                limit_key="kubelet_api_query",
-            )
-            raise KubeletApiException(
-                "Invalid response from Kubelet API when querying '%s': %s"
-                % (path, six.text_type(response))
-            )
+            if not self._fallback:
```

Comment:
Is there a particular status code we can key off of to trigger the fallback?  It seems here that any error (such as us issuing a poorly formatted request) will trigger us to fallback.  That worries me a bit.

Reply:
I realized there was no need for the exception check since thats not the response using the wrong port gives, I didn't follow the suggestions exactly as a result since I think I can make it cleaner now.